runtime/src/accounts.rs

```diff
@@ -292,20 +322,40 @@ impl Accounts {
     }
 
     fn lock_account(
-        locks: &mut HashSet<Pubkey>,
+        (fork_locks, parent_locks): &mut AccountLocks,
         keys: &[Pubkey],
         error_counters: &mut ErrorCounters,
     ) -> Result<()> {
         // Copy all the accounts
+        let mut fork_locks = fork_locks.lock().unwrap();
         for k in keys {
-            if locks.contains(k) {
+            let is_locked = {
+                if fork_locks.contains(k) {
+                    true
+                } else {
+                    // Check parent locks. As soon as a set of parent locks is empty,
+                    // we can remove it from the list b/c that means the parent has
+                    // released the locks.
+                    let mut is_locked = false;
+                    parent_locks.retain(|p| {
```

Comment:
this looks a lot like the really slow load_account code..

Reply:
It should be faster in that rarely should there be more than one parent (when a parent contains accounts that have not yet been unlocked on a slot boundary), and when that parent's locks are cleaned up, the parent is removed from the child.